api/filters/replacement/replacement.go

```diff
@@ -118,22 +118,33 @@ func applyToNode(node *yaml.RNode, value *yaml.RNode, target *types.TargetSelect
 		fieldPath := kyaml_utils.SmarterPathSplitter(fp, ".")
 		var t *yaml.RNode
 		var err error
+		var create bool
+
 		if target.Options != nil && target.Options.Create {
 			// create option is not supported in a wildcard matching.
 			// Because, PathMatcher is not supported create option.
-			// So, if create option is set, we fallback to PathGetter.
+			// So, if create option is set, we fall back to PathGetter.
 			for _, f := range fieldPath {
 				if f == "*" {
 					return errors.New("cannot support create option in a multi-value target") //nolint:goerr113
 				}
 			}
+			found, err := node.Pipe(yaml.Lookup(fieldPath...))
+			if err != nil {
+				return err
+			}
+			create = found == nil
+		}
+
+		if create {
 			t, err = node.Pipe(yaml.LookupCreate(value.YNode().Kind, fieldPath...))
 		} else {
 			t, err = node.Pipe(&yaml.PathMatcher{Path: fieldPath})
 		}
 		if err != nil {
 			return err
 		}
+
```

Comment:
I know in https://github.com/kubernetes-sigs/kustomize/pull/4574#discussion_r846466912, we discussed refactoring this very confusing code, but I'd prefer to do the refactor in a followup PR if needed.

Reply:
I would really prefer it get done in this PR, since it further complicates the code path and the ease of understanding what's going on here.

I see your solution is to bypass creation when we don't really need it, which is fair. But do we have a bug in `LookupCreate` we should be fixing? Should the "already exists" bypass go in there?

Reply:
> I would really prefer it get done in this PR

I refactored the code to make it more readable. If there any parts that you still find hard to read, I can further try to clean that up

> I see your solution is to bypass creation when we don't really need it, which is fair. But do we have a bug in LookupCreate we should be fixing? Should the "already exists" bypass go in there?

I spent a few hours today trying to do this, an in practice I agree that the "already exists" bypass should be as deep as possible, but I found it surprisingly difficult (largely because I'm not that familiar with how all these filters work). It appears that `LookupCreate` uses the `PathGetter` filter (which does not support the new wildcard option), while we are using the `PathMatcher` for creation. 

`PathGetter` and `PathMatcher` seem to differ in nuanced ways - one will wrap the result in a Sequence node while the other doesn't; sometimes in `PathGetter` the pointer for the RNode is one node off from what I expect it to be, etc. If you have a good understanding of the code and know a good place to move the "bypass" let me know. I'll look into it further myself and I'm happy with the principle of it; but after today I feel that it may be more trouble than it's worth.

I moved the "bypass" to its own helper function that checks whether or not we need to create the node in the meantime, and I hope that that can be enough to move forward here.

Reply:
I still find two things confusing tbh. One is just naming nitpicks... the words "target" (target selector or actual target resource?) and "node" (entire resource, field, from what?) are ambiguous in this context, and sometimes the pluralization seems incorrect (`copyValueToTargets` is only passed one specific target).

The second is the bit around the case statement below. It knows things about the structure of the nodes it deals with that really come from the implementation differences between LookupCreate and PathMatcher. I.e. I cant' tell from the code when/why we end up with an empty YNode, and why we otherwise need to dig through the contents. I [created a branch](https://github.com/kubernetes-sigs/kustomize/commit/d5bafe9ccc3d45a5ffc01422e0c2efe9aee1ac77#diff-7af4e50a78ba8c17345cb7e71ee831cb2f0ff2299480eb71f6188e89c312f298R180) with a suggestion (including a bunch of naming nits you can take or leave).

Re: whether we have a bug in LookupCreate, I didn't mean moving the bypass so much as identifying why it is necessary to begin with, and potentially fixing that. I would expect LookupCreate to be able to successfully no-op. I wrote a likely redundant [regression test](https://github.com/kubernetes-sigs/kustomize/commit/d5bafe9ccc3d45a5ffc01422e0c2efe9aee1ac77#diff-b4b4edfb62aa52f449c024780fab248bc029ccad387ac9d82ceb66e34cfdab13R584) on that level to prove LookupCreate is at fault... and it isn't! So, it's something in the replacements code. What?

Still on my [suggestion branch](https://github.com/kubernetes-sigs/kustomize/commit/d5bafe9ccc3d45a5ffc01422e0c2efe9aee1ac77#diff-7af4e50a78ba8c17345cb7e71ee831cb2f0ff2299480eb71f6188e89c312f298R180), I tried removing the bypass to start debugging this... and discovered that on my branch, it was already fixed! It seems to me the problem is with the switch statement. If we came from LookupCreate, and the node already existed, we have a non-sequence root that has non-zero content. Instead of appending it, we iterate over its contents and append each piece--oops!

Reply:
> Still on my suggestion branch, I tried removing the bypass to start debugging this... and discovered that on my branch, it was already fixed! It seems to me the problem is with the switch statement. If we came from LookupCreate, and the node already existed, we have a non-sequence root that has non-zero content. Instead of appending it, we iterate over its contents and append each piece--oops!

🤦‍♀️ Thank you for digging in and finding that.

I pulled your suggestion branch into this PR, I like how you refactored the code and renamed things.